Widgets_CDAT/DictWidget/dict.py

```diff
@@ -0,0 +1,219 @@
+        cur_text = cur_box.itemText(cur_box.currentIndex())
+        selectedItems = []
+        for box in self.cBoxes:
+            check_text = box.itemText(box.currentIndex())
+
+            selectedItems.append(check_text)
+
+            # check if boxes match and set to blank if so
+            if box != cur_box:
+                if check_text == cur_text:
+                    box.setCurrentIndex(0)
+
+        return selectedItems
+
+
+    # updates dict based on lineedit values
+    def updateEditTexts(self):
+        texts = []
+        for item in self.textEdits:
+            texts.append(item.text())
+        for i in self.validators:
+            i.texts = self.keys
+
+
+    # update lists and selectedItems
+    def updateLists(self, left, right, index=None):
+        if self.valid_keys:
+            selectedItems = self.updateCBoxes(left)
+            self.keys = selectedItems
+            index = self.keys.index(left.itemText(left.currentIndex()))
+            self.values[index] = right.text()
+
+
+        else:
+            key = left.text()
+            value = right.text()
+            index = self.textEdits.index(left)
+            self.keys[index] = key
+            self.values[index] = value
+            self.updateEditTexts()
+
+        self.textEdited.emit(dict(zip(self.keys, self.values)))
+
+
+    # populate if dictionary is given
+    def insertRow(self, key="", value=""):
+
+        if self.valid_keys and self.row_count >= len(self.valid_keys)-1:
+            return
+
+        self.keys.append(key)
+        self.values.append(value)
+
+        # for key, value in self.dict.items():
+        col = QHBoxLayout()
+        r_button = QPushButton()
+        r_button.setText("X")
+        r_button.clicked.connect(partial(self.removeRow, col, self.row_count))
+
+        # create right text edit
+        edit_right = QLineEdit()
+        edit_right.setText(str(value))
+        edit_right.setMinimumWidth(100)
+
+        # set text based off if valid_keys has values
+        if self.valid_keys:
+            edit_left = QComboBox()
+            for i in self.valid_keys:
+                edit_left.addItem(i)
+            index = edit_left.findText(str(key))
+            if index != -1:
+                edit_left.setCurrentIndex(index)
+
+            # assign signals
+            edit_left.currentIndexChanged.connect(partial(self.updateLists, edit_left, edit_right))
+
+            self.cBoxes.append(edit_left)
+
+        else:
+            edit_left = QLineEdit()
+            validator = InputChecker(self.row_count)
+            self.validators.append(validator)
+            edit_left.setValidator(validator)
+            validator.correctInput.connect(partial(self.colorValid, edit_left))
+            validator.inputInvalid.connect(partial(self.colorInvalid, edit_left))
+            edit_left.setText(str(key))
+
+            edit_left.editingFinished.connect(partial(self.updateLists, edit_left, edit_right))
```

Comment:
Since this, line 130, and line 147 all use a partial made with the same arguments, you should probably just make one above and use it in all three places.